docs/featsel.1
.\" Process this file with
.\" groff -man -Tascii featsel.1
.\"
.TH FEATSEL 1 "MARCH 2010" Linux "User Manuals"
.SH NAME
featsel \- a flexible feature selection program
.SH SYNOPSIS
.B featsel [-h] [-f
.I input-file-name
.B ]
.B [ -c 
.I cost-function-name
.B ] ...
.SH DESCRIPTION
.B featsel
receives a set
.B S 
of objects, values associated with each 
of them, a cost function 
.B c
with subsets of S as domain
and real numbers as image, and searches for a subset 
X in S such as c(X) is minimum (i.e., there are no 
Y in S, such as c(Y) < c(X)). 
The minima subsets found are printed into the standard
output.
The featsel program may use as search algorithm an
implementation of the U-Curve-Search (UCS), and as the
cost function the hamming distance and the absolute sum.

.SH OPTIONS
.IP "-a algorithm-name"
Selects the
.I algorithm-name
as the algorithm used in the search. The default value is 
.I ucs
(U-Curve Search); other valid names are
.I ubb
(U-Curve Branch and Bound),
.I pfs
(Poset Forest Search),
.I sffs
(Sequential Forward Floating Selection),
.I sfs
(Sequential Forward Selection), and
.I es
(Exhaustive Search).
.IP "-c cost-function-name"
Selects the
.I cost-function-name
as the cost function of the search. The default value is 
.I abs_sum;
other valid names are 
.I hamming_distance,
.I mce 
(Mean Conditional Entropy for W-operator design),
.I explicit 
(cost function values explicitly defined from file),
.I mean_abs_sum.
.IP "-f input-file-name"
Uses as instance either a XML or a DAT file 
.I input-file-name,
containing the elements of the set, their names and values.
.IP -h
Displays this help screen.
.IP "-m number-of-minima"
The number of minima to be showed in the standard output. It
should be and integer grater than zero, and the default value
is 1.
.IP "-n number-of-elements"
If the option -f is not used, creates a random set with
.I number-of-elements
elements with a single value. It should be an integer
greater than zero. The default value is 3.
.IP "-r range"
If the option -f is not used, creates the elements of
the random set with a single value greater or equal than
zero and less than
.I range.
The default value is 1000.
.IP -s
Stores all the visited subsets into the file called
.I visited.log
.IP "-t1 threshold"
Set a
.I threshold
value that ends the execution if a value less or equal to
it is computed by the cost function. Used in the first step
(pre-processing) of a heuristic procedure.
.IP "-t2 max-number-of-calls-of-cost-function"
Set a
.I max-number-of-calls-of-cost-function
value that ends the execution if the cost function is computed
that number of times. It is used in the second step of a
heuristic procedure.


.SH FILES
.I etc/featsel.conf
.RS
The program configuration file (not implemented yet).
.RE
.I etc/costfunc.conf
.RS
Configuration of the cost function lexical/syntactic analyzer (not implemented yet).
.SH ENVIRONMENT
.IP PATH
May be useful to add the full pathname of the directory where the featsel binary
is located. The configuration files must be at the same directory of the binary.
.SH DIAGNOSTICS
The following diagnostics may be issued on stderr:
.RE
Bad magic number.
.RS
The input file does not look like a XML file.
.RE
Bad cost.
.RS
.B "cost-function-name"
given is not a valid one.
.SH BUGS
XML parsing should be improved.
.SH AUTHOR
Marcelo S. Reis < msreis at ime dot usp dot br>
.SH "SEE ALSO"
.BR featsel (1)


src/featsel.cpp
//============================================================================
//
//    featsel -- a flexible feature selection program.
//    Copyright (C) 2010  Marcelo S. Reis
//
//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//============================================================================


#include "global.h"
#include "string.h"
#include "ElementSet.h"
#include "CostFunction.h"

#include "functions/AbsSum.h"
#include "functions/Explicit.h"
#include "functions/HammingDistance.h"
#include "functions/MeanAbsSum.h"
#include "functions/MeanConditionalEntropy.h"

#include "algorithms/UCurveSearch.h"
#include "algorithms/ExhaustiveSearch.h"
#include "algorithms/SFS.h"
#include "algorithms/SFFS.h"
#include "algorithms/BranchAndBound.h"
#include "algorithms/PosetForestSearch.h"


// Function to parse and verify the correctness of the parameters.
// It returns '0' if all the parameters were correct, '1' if there was an error,
// and '2' if the option '-h' (help) was called.
//
int parse_parameters
    (int, char **, string *, unsigned int *, string *,
     unsigned int *, unsigned int *, bool *, string *, float *, unsigned int *);


// The main function.
//
int main(int argc, char * argv[])
{
	unsigned int max_number_of_minima = 1;
	unsigned int n = 3;
	unsigned int range = 1000;
	unsigned int max_number_of_calls_of_cost_function = 0; // if == 0, then there is no limit
	int i;
	CostFunction * cost_function;
	Solver * solver;
	ElementSet * S;
	string list;
	string a_cost_function ("abs_sum");
	string algorithm ("ucs");
	string file_name;
	bool store_visited_subsets = false;
	float threshold = FLT_MIN;
	ofstream log_file;

	// parses the parameters
	i = parse_parameters(argc, argv, &file_name, &max_number_of_minima, &a_cost_function,
			             &n, &range, &store_visited_subsets, &algorithm, &threshold,
			             &max_number_of_calls_of_cost_function);

	if (i != EXIT_SUCCESS)    // help or error in parameters
		return EXIT_FAILURE;
	else
	{
		// selects the type of instance (from XML file, from DAT file, or random)
		//
		if ((file_name.size () > 0) &&
			((file_name.find (".xml") != string::npos) || (file_name.find (".XML") != string::npos)))
			S = new ElementSet ("S", file_name.data ());
		else if ((file_name.size () > 0) &&
				((file_name.find (".dat") != string::npos) || (file_name.find (".DAT") != string::npos)))
		{
		    // IMPORTANT: the use of .dat file requires the -n option!
			S = new ElementSet (file_name.data (), n);
		}
		else
			S = new ElementSet ("S", n, range);

		// Selects the proper solver
		//
		if (algorithm.compare ("ucs") == 0)
			solver = new UCurveSearch ();
		else if (algorithm.compare ("es") == 0)
			solver = new ExhaustiveSearch ();
		else if (algorithm.compare ("sfs") == 0)
			solver = new SFS ();
		else if (algorithm.compare ("sffs") == 0)
		{
			solver = new SFFS ();
			// S->permute_set ();    // TODO: shuffle the set to obtain a different running time
		}
		else if (algorithm.compare ("pfs") == 0)
			solver = new PosetForestSearch ();
		else if (algorithm.compare ("ubb") == 0)
		{
			solver = new BranchAndBound ();
			// S->permute_set ();    // TODO: shuffle the set to obtain a different running time
		}
		else
			return EXIT_FAILURE;

		// Selects the proper cost function
		//
		if (a_cost_function.compare ("mean_abs_sum") == 0)
			cost_function = new MeanAbsSum (S);
		else if (a_cost_function.compare ("abs_sum") == 0)
			cost_function = new AbsSum (S);
		else if (a_cost_function.compare ("explicit") == 0)
			cost_function = new Explicit (S);
		else if (a_cost_function.compare ("hamming_distance") == 0)
			cost_function = new HammingDistance (S);
		else if (a_cost_function.compare ("mce") == 0)
			cost_function = new MeanConditionalEntropy (S);
		else
			return EXIT_FAILURE;

		// If the solver has a heuristic strategy, here we
		// set the threshold value for the cost function,
		// for the first stage (pre-processing).
		//
		if (threshold > FLT_MIN)
			cost_function->set_threshold (threshold);

		// If the solver has a heuristic strategy, here we
		// set the maximum number of calls of the cost function,
		// for the second stage (search with limited range).
		//
		if (max_number_of_calls_of_cost_function > 0)
			cost_function->set_max_number_of_calls_of_cost_function
			                  (max_number_of_calls_of_cost_function);

		// Sets the solver parameters
		//
		solver->set_parameters (cost_function, S, store_visited_subsets);

		// Runs the solver
		//
		solver->get_minima_list (max_number_of_minima);

		// Prints the results
		//
		cout << endl << "== List of best subsets found ==" << endl;
		cout << " X : c(X)" << endl;
		cout << solver->print_search_report ();
		if (store_visited_subsets)
		{
			cout << endl << "Visited subsets stored into 'visited.log' file" << endl;
			log_file.open ("visited.log");
			log_file << solver->print_list_of_visited_subsets () << endl;
			log_file.close ();
		}
		else
		{
			int elapsed_time =
					cost_function->get_the_elapsed_time_of_the_calls_of_this_cost_function ();
			// number of visited subsets is the number of calls
			// of the cost function!
			int number_of_subsets = solver->print_number_of_visited_subsets ();
			if (number_of_subsets > 0)
			{
				cout << endl << "Number of visited subsets: "
				 <<  number_of_subsets << endl;
				cout << "Required time to compute the visited subsets: "
				 << elapsed_time << " microseconds " << endl;
				cout << "(average " << elapsed_time / number_of_subsets
				 << " microseconds per node)" << endl;
			}
			cout << endl << "Elapsed time of execution of the algorithm (in microseconds): "
				 << solver->get_elapsed_time_of_the_algorithm () << endl;

			if ( (algorithm.compare ("ucs") == 0)
			   )
			{
				cout << endl << "Elapsed time of the minima exhausting (in microseconds): "
				     << solver->get_elapsed_time_of_all_calls_of_the_minima_exhausting () << endl;
				cout << endl << "Number of calls of the minima exhausting: "
					 << solver->get_number_of_calls_of_minimum_exhausting () << endl;
				cout << endl << "Maximum size a graph achives during the minima exhausting: "
					 << solver->get_max_size_of_the_graph () << endl;
				cout << endl << "Number of times a minimal (maximal) element is computed: "
					 << solver->get_number_of_calls_of_minimal_and_maximal_element () << endl;
			}

			if (algorithm.compare ("pfs") == 0)
			{
				cout << endl << "Number of iterations of the main algorithm: "
					 << solver->get_number_of_iterations () << endl;
				cout << endl << "Maximum size that achieves forest A: "
					 << solver->get_max_size_of_the_forest_A () << endl;
				cout << endl << "Maximum size that achieves forest B: "
					 << solver->get_max_size_of_the_forest_B () << endl;
			}
		}

		cout << endl << "== End of processing ==" << endl << endl;

		// exits to the OS
		//
        delete S;
        delete solver;
        delete cost_function;
		return EXIT_SUCCESS;
	}
} // end of main



// function implementation
//
int parse_parameters (int argc, char ** argv, string * file_name,
	                  unsigned int * m, string * c, unsigned int * n,
	                  unsigned int * range, bool * keep_subsets,
	                  string * a, float * threshold,
	                  unsigned int * max_number_of_calls_of_cost_function)
{
	int i;
	bool error = false;
	string a_line;
	ifstream a_file;
	string disclaimer ("featsel -- a flexible feature selection program.\n\
Copyright (C) 2010  Marcelo S. Reis.\n\n \
This program comes with ABSOLUTELY NO WARRANTY.\n \
This is free software, and you are welcome to redistribute it\n \
under certain conditions; see 'LICENSE.TXT' for details.");
	// parsing the arguments
	//
	for (i=1; i<argc; i++)
	{
		// -f
		//
		if( (strcmp (argv[i],"-f") == 0) && ((i+1) < argc) )
		{
			a_file.open (argv[++i]);
			if (a_file.is_open ())
			{
				a_file.close ();
				file_name->append (argv[i]);
			}
			else
			{
				cout << "\nError: unable to open input file '" << argv[i] << "'!" << endl;
				error = true;
			}
		}
		else if ( (strcmp (argv[i],"-f") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-f' must have a file name." << endl;
			error = true;
		}
		// -r
		//
		else if ( (strcmp( argv[i],"-r") == 0) && ((i+1) < argc) )
		{
			*range = atoi (argv[++i]);
			if (*range >= RAND_MAX)
			{
				cout << "\nError: range should be in [0, " << RAND_MAX << ") !\n";
				error = true;
			}
		}
		else if ( (strcmp (argv[i],"-r") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-r' must have a value." << endl;
			error = true;
		}
		// -t1  pre-processing of the heuristic mode; it receives a threshold
		//
		else if ( (strcmp( argv[i],"-t1") == 0) && ((i+1) < argc) )
		{
			*threshold = atof (argv[++i]);
		}
		else if ( (strcmp (argv[i],"-t1") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-t1' must have a value." << endl;
			error = true;
		}
		// -t2  second stage of the heuristic mode; it receives a limit to
		//      the number of calls of the cost function
		//
		else if ( (strcmp( argv[i],"-t2") == 0) && ((i+1) < argc) )
		{
			*max_number_of_calls_of_cost_function = atoi (argv[++i]);
		}
		else if ( (strcmp (argv[i],"-t2") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-t2' must have a value." << endl;
			error = true;
		}
		// -n
		//
		else if ( (strcmp (argv[i],"-n") == 0) && ((i+1) < argc) )
		{
			if (atoi (argv[++i]) > 0)
				*n = atoi (argv[i]);
			else
			{
				cout << "\nError: number of elements should be a positive integer!" << endl;
				error = true;
			}
		}
		else if ( (strcmp (argv[i],"-n") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-n' must have a value." << endl;
			error = true;
		}
		// -c
		//
		else if ( (strcmp (argv[i],"-c") == 0) && ((i+1) < argc) )
		{
			i++;
			if ( (strcmp (argv[i], "abs_sum") == 0) ||
			 	 (strcmp (argv[i], "mean_abs_sum") == 0) ||
				 (strcmp (argv[i], "mce") == 0) ||
				 (strcmp (argv[i], "explicit") == 0) ||
				 (strcmp (argv[i], "hamming_distance") == 0) )
			{
				c->clear();
				c->append(argv[i]);
			}
			else
			{
				cout << "\nError: unknown cost function '" << argv[i] << "'!" << endl;
				error = true;
			}
		}
		// -a
		//
		else if ( (strcmp(argv[i],"-a") == 0) && ((i+1) < argc) )
		{
			i++;
			if ( (strcmp(argv[i], "ucs") == 0) ||
			     (strcmp(argv[i], "es") == 0) ||
			     (strcmp(argv[i], "sfs") == 0) ||
			     (strcmp(argv[i], "sffs") == 0) ||
			     (strcmp(argv[i], "pfs") == 0) ||
			     (strcmp(argv[i], "ubb") == 0) )
			{
				a->clear();
				a->append(argv[i]);
			}
			else
			{
				cout << "\nError: unknown algorithm '" << argv[i] << "'!" << endl;
				error = true;
			}
		}
		else if ( (strcmp(argv[i],"-c") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-c' must have a cost function name." << endl;
			error = true;
		}
		// -m
		//
		else if ( (strcmp (argv[i],"-m") == 0) && ((i+1) < argc) )
		{
			if (atoi (argv[++i]) > 0)
				*m = atoi (argv[i]);
			else
			{
				cout << "\nError: maximum number of minima should be a positive integer!" << endl;
				error = true;
			}
		}
		else if ( (strcmp (argv[i],"-m") == 0) && ((i+1) >= argc) )
		{
			cout << "\nError: parameter '-m' must have a value." << endl;
			error = true;
		}
		// -s
		//
		else if (strcmp (argv[i],"-s") == 0)
		{
			*keep_subsets = true;
		}
		// -h
		//
		else if (strcmp (argv[i],"-h") == 0)
		{
			cout << endl << disclaimer << endl << endl;
			a_file.open ("featsel.txt");
			if(a_file.is_open ())
				while(a_file.good ())
				{
					getline (a_file, a_line);
					cout << a_line << endl;
				}
			a_file.close ();
			return EXIT_FAILURE;
		}
		else
		{
			cout <<  endl << "Unknown option '" << argv[i] << "'." << endl;
			error = true;
		}
	}
	if ( (error) || (argc == 1) )
	{
		cout << endl << disclaimer << endl;
		cout << endl << "Type 'featsel -h' for help." << endl << endl;
		return EXIT_FAILURE;
	}
	else
		return EXIT_SUCCESS;

} // end of function

